File: apps/api/schemas/workspace.py
"""
Pydantic schemas for workspace data.

These mirror the shared TypeScript types and are used for
request/response validation on the API side.
"""

from enum import Enum

from pydantic import BaseModel, Field


class ComponentType(str, Enum):
    timeline = "timeline"
    task = "task"
    note = "note"


class Priority(str, Enum):
    urgent = "urgent"
    high = "high"
    normal = "normal"
    low = "low"


class WorkspaceComponent(BaseModel):
    id: str
    type: ComponentType
    title: str
    description: str = ""
    priority: Priority = Priority.normal
    date: str | None = None
    time_slot: str | None = Field(None, alias="timeSlot")
    completed: bool = False
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")

    model_config = {"populate_by_name": True}


class WorkspaceMeta(BaseModel):
    active_id: str | None = Field(None, alias="activeId")
    last_entities: dict[str, str] = Field(default_factory=dict, alias="lastEntities")
    session_id: str = Field(alias="sessionId")

    model_config = {"populate_by_name": True}

File: packages/shared/src/types/workspace.ts
/**
 * Unique identifier for workspace components.
 * Generated client-side using nanoid for offline-first creation.
 */
export type ComponentId = string;

/**
 * The types of visual components that can exist in a workspace.
 * Kept minimal for reliability -- extend only when stable.
 */
export type ComponentType = "timeline" | "task" | "note";

/**
 * Priority levels for tasks. "normal" is the default.
 */
export type Priority = "urgent" | "high" | "normal" | "low";

/**
 * A single component in the workspace.
 * This is the source of truth stored in the Yjs CRDT document.
 */
export interface WorkspaceComponent {
  id: ComponentId;
  type: ComponentType;
  title: string;
  description?: string;
  priority: Priority;
  date?: string;
  timeSlot?: string;
  completed: boolean;
  createdAt: string;
  updatedAt: string;
}

/**
 * Metadata about the current workspace session.
 * Used for entity resolution and UI state.
 */
export interface WorkspaceMeta {
  activeId: ComponentId | null;
  lastEntities: Record<string, ComponentId>;
  sessionId: string;
}

/**
 * The full workspace document shape.
 * Maps directly to the Yjs shared types structure.
 */
export interface WorkspaceDocument {
  components: Record<ComponentId, WorkspaceComponent>;
  order: ComponentId[];
  meta: WorkspaceMeta;
}
